flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManagerDriver.java

```diff
@@ -260,36 +257,47 @@ public void releaseResource(YarnWorkerNode workerNode) {
 	//  Internal
 	// ------------------------------------------------------------------------
 
-	private void onContainersOfResourceAllocated(Resource resource, List<Container> containers) {
-		final List<TaskExecutorProcessSpec> pendingTaskExecutorProcessSpecs =
-			taskExecutorProcessSpecContainerResourceAdapter.getTaskExecutorProcessSpec(resource, matchingStrategy).stream()
-				.flatMap(spec -> Collections.nCopies(getNumRequestedNotAllocatedWorkersFor(spec), spec).stream())
-				.collect(Collectors.toList());
+	private void onContainersOfPriorityAllocated(Priority priority, List<Container> containers) {
+		final Optional<TaskExecutorProcessSpec> taskExecutorProcessSpecOpt =
+			taskExecutorProcessSpecContainerResourcePriorityAdapter.getTaskExecutorProcessSpec(priority);
+
+		if (!taskExecutorProcessSpecOpt.isPresent()) {
+			log.warn("Receive {} containers with unrecognized priority {}. This should not happen.",
+				containers.size(), priority.getPriority());
+			for (Container container : containers) {
+				returnExcessContainer(container);
+			}
+			return;
+		}
 
-		int numPending = pendingTaskExecutorProcessSpecs.size();
-		log.info("Received {} containers with resource {}, {} pending container requests.",
+		final TaskExecutorProcessSpec taskExecutorProcessSpec = taskExecutorProcessSpecOpt.get();
+
+		final Optional<Resource> resourceOpt = taskExecutorProcessSpecContainerResourcePriorityAdapter.getResource(taskExecutorProcessSpec);
+		Preconditions.checkState(resourceOpt.isPresent());
```

Comment:
If the taskExecutorProcessSpecOpt is present, the resourceOpt should be present by design. I think we may have `taskExecutorProcessSpecContainerResourcePriorityAdapter#getTaskExecutorProcessSpecAndResource`, directly return an optional `Tuple2`. So, we do not need to call `taskExecutorProcessSpecContainerResourcePriorityAdapter.getResource(taskExecutorProcessSpec)` and check whether resourceOpt is present.

Reply:
We could also make the value of `priorityToTaskExecutorProcessSpec` to a `Tuple2` of `TaskExecutorProcessSpec` and `Resource`.